crates/sui-source-validation-service/src/lib.rs

```diff
@@ -61,16 +64,108 @@ pub fn parse_config(config_path: impl AsRef<Path>) -> anyhow::Result<Config> {
 }
 
-pub async fn clone_repositories(config: &Config) -> anyhow::Result<()> {
+#[derive(Debug)]
+/// Represents a sequence of git commands to clone a repository and sparsely checkout Move packages within.
+pub struct CloneCommand {
+    /// git args
+    args: Vec<Vec<String>>,
+    /// report repository url in error messages
+    repo_url: String,
+}
+
+impl CloneCommand {
+    pub fn new(p: &Packages, dest: &Path) -> anyhow::Result<CloneCommand> {
+        let repo_url = Url::parse(&p.repository)?;
+        let Some(components) = repo_url.path_segments().map(|c| c.collect::<Vec<_>>()) else {
+	    bail!("Could not discover repository path in url {}", &p.repository)
+	};
+        let Some(repo_name) = components.last() else {
+	    bail!("Could not discover repository name in url {}", &p.repository)
+	};
+        let dest = dest
+            .join(repo_name)
+            .into_os_string()
+            .into_string()
+            .map_err(|_| {
+                anyhow!(
+                    "Could not create path to clone repsository {}",
+                    &p.repository
+                )
+            })?;
+
+        let mut args = vec![];
+        // Args to clone empty repository
+        args.push(
+            [
+                "clone",
+                "-n",
+                "--depth=1",
+                "--filter=tree:0",
+                &p.repository,
+                &dest,
+            ]
+            .iter()
+            .map(|&s| s.into())
+            .collect(),
+        );
+
+        // Args to sparse checkout the package set
+        let mut prefix: Vec<String> = ["-C", &dest, "sparse-checkout", "set", "--no-cone"]
+            .iter()
+            .map(|&s| s.into())
+            .collect();
+        prefix.extend_from_slice(&p.paths);
+        args.push(prefix);
+
+        // Args to checkout the default branch.
+        args.push(
+            ["-C", &dest, "checkout"]
+                .iter()
+                .map(|&s| s.into())
+                .collect(),
+        );
+
+        Ok(Self {
+            args,
+            repo_url: p.repository.clone(),
+        })
+    }
+
+    pub async fn run(&self) -> anyhow::Result<()> {
+        for args in &self.args {
+            Command::new("git").args(args).output().map_err(|_| {
+                anyhow!(
+                    "Error cloning package(s) for {} with command git {:#?}",
+                    self.repo_url,
+                    args
+                )
+            })?;
+        }
+        Ok(())
+    }
+}
+
+/// Clones repositories and checks out packages as per `config` at the directory `dir`.
+pub async fn clone_repositories(config: &Config, dir: &Path) -> anyhow::Result<()> {
+    let mut tasks = vec![];
     for p in &config.packages {
-        let _ = p.repository;
-        let _ = p.paths;
+        let command = CloneCommand::new(p, dir)?;
+        let t = tokio::spawn(async move { command.run().await });
```

Comment:
Is this going to do something funky if two packages end up mapping to the same directory?

Reply:
Shouldn't cause something funky--we're just using the package paths in the config to checkout on the directory structure of the repo, so you can have, e.g., subpaths or specify the same path multiple times. 

One thing `sparse-checkout` (and we) don't guarantee yet is that those directory paths actually exist after checking out. This is something we can validate after cloning, and warn/error if we find a package path in the config that doesn't match up to a path in the repo.